File: src/event/Touch.ts
import { $canceled } from './private/Touch'
import { $captured } from './private/Touch'
import { $id } from './private/Touch'
import { $target } from './private/Touch'
import { $x } from './private/Touch'
import { $y } from './private/Touch'
import { View } from '../view/View'

/**
 * @class Touch
 * @since 0.1.0
 */
export class Touch {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property x
	 * @since 0.1.0
	 */
	public get x(): number {
		return this[$x]
	}

	/**
	 * @property y
	 * @since 0.1.0
	 */
	public get y(): number {
		return this[$y]
	}

	/**
	 * @property id
	 * @since 0.7.0
	 */
	public get id(): number {
		return this[$id]
	}

	/**
	 * @property target
	 * @since 0.1.0
	 */
	public get target(): View {
		return this[$target]
	}

	/**
	 * @property captured
	 * @since 0.7.0
	 */
	public get captured(): boolean {
		return this[$captured]
	}

	/**
	 * @property canceled
	 * @since 0.7.0
	 */
	public get canceled(): boolean {
		return this[$canceled]
	}

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.1.0
	 */
	constructor(target: View) {
		this[$target] = target
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	/**
	 * @property $x
	 * @since 0.7.0
	 * @hidden
	 */
	private [$x]: number = 0

	/**
	 * @property $y
	 * @since 0.7.0
	 * @hidden
	 */
	private [$y]: number = 0

	/**
	 * @property $id
	 * @since 0.7.0
	 * @hidden
	 */
	private [$id]: number = 0

	/**
	 * @property $target
	 * @since 0.7.0
	 * @hidden
	 */
	private [$target]: View

	/**
	 * @property $canceled
	 * @since 0.7.0
	 * @hidden
	 */
	private [$canceled]: boolean = false

	/**
	 * @property $captured
	 * @since 0.7.0
	 * @hidden
	 */
	private [$captured]: boolean = false
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/application/touch/Touch.kt
package ca.logaritm.dezel.application.touch

import android.view.MotionEvent
import ca.logaritm.dezel.extension.view.pointer
import ca.logaritm.dezel.modules.view.JavaScriptView

/**
 * @class Touch
 * @since 0.7.0
 */
public class Touch(pointer: Int, target: JavaScriptView) {

	//--------------------------------------------------------------------------
	// Static
	//--------------------------------------------------------------------------

	companion object {

		/**
		 * @property touches
		 * @since 0.7.0
		 * @hidden
		 */
		private var touches: MutableMap<Int, Touch> = mutableMapOf()

		/**
		 * @method get
		 * @since 0.7.0
		 */
		public fun get(event: MotionEvent): Touch? {
			return this.touches[event.pointer]
		}

		/**
		 * @method acquire
		 * @since 0.7.0
		 */
		public fun acquire(event: MotionEvent, target: JavaScriptView): Touch {

			if (this.touches[event.pointer] == null) {
				this.touches[event.pointer] = Touch(event.pointer, target)
				return this.touches[event.pointer]!!
			}

			throw Exception("Unexpected error.")
		}

		/**
		 * @method release
		 * @since 0.7.0
		 */
		public fun release(event: MotionEvent) {
			this.touches.remove(event.pointer)
		}
	}

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property pointer
	 * @since 0.7.0
	 */
	public var id: Int

	/**
	 * @property target
	 * @since 0.7.0
	 */
	public var target: JavaScriptView

	/**
	 * @property x
	 * @since 0.7.0
	 */
	public var x: Double = 0.0

	/**
	 * @property y
	 * @since 0.7.0
	 */
	public var y: Double = 0.0

	/**
	 * @property canceled
	 * @since 0.7.0
	 */
	public var canceled: Boolean = false

	/**
	 * @property captured
	 * @since 0.7.0
	 */
	public var captured: Boolean = false

	/**
	 * @property disposed
	 * @since 0.7.0
	 */
	public var disposed: Boolean = false

	/**
	 * @property receiver
	 * @since 0.7.0
	 */
	public var receiver: JavaScriptView? = null

	/**
	 * @constructor
	 * @since 0.7.0
	 */
	init {
		this.id = pointer
		this.target = target
	}
}